contracts/middleware/MiddlewareRemove.sol

```diff
@@ -0,0 +1,128 @@
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+contract MiddlewareRemove is BaseRemove {
+    using CustomRevert for bytes4;
+    using Hooks for IHooks;
+    using TransientStateLibrary for IPoolManager;
+
+    error HookModifiedDeltasBeforeRemove();
+    error HookTookTooMuchFee();
+    error HookInvalidDeltasAfterRemove();
+    error MaxFeeBipsTooHigh();
+
+    constructor(IPoolManager _manager, address _impl, uint256 _maxFeeBips) BaseRemove(_manager, _impl) {
+        if (_maxFeeBips > MAX_BIPS) revert MaxFeeBipsTooHigh();
+        maxFeeBips = _maxFeeBips;
+    }
+
+    function afterRemoveLiquidity(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external returns (bytes4, BalanceDelta) {
+        if (bytes32(hookData) == OVERRIDE_BYTES) {
+            (, bytes memory returnData) = address(implementation).delegatecall(
+                abi.encodeWithSelector(this.afterRemoveLiquidity.selector, sender, key, params, delta, hookData[32:])
+            );
+            return abi.decode(returnData, (bytes4, BalanceDelta));
+        }
+        (bool success, bytes memory returnData) = address(this).delegatecall{gas: GAS_LIMIT}(
+            abi.encodeWithSelector(this._callAndEnsureValidDeltas.selector, sender, key, params, delta, hookData)
+        );
+        if (success) {
+            return (BaseHook.afterRemoveLiquidity.selector, abi.decode(returnData, (BalanceDelta)));
+        } else {
+            return (BaseHook.afterRemoveLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+        }
+    }
+
+    function _callAndEnsureValidDeltas(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external returns (BalanceDelta) {
+        (bool success, bytes memory returnData) = address(implementation).delegatecall(
+            abi.encodeWithSelector(this.afterRemoveLiquidity.selector, sender, key, params, delta, hookData)
+        );
+        if (!success) {
+            revert FailedImplementationCall();
+        }
+        (bytes4 selector, BalanceDelta returnDelta) = abi.decode(returnData, (bytes4, BalanceDelta));
+        if (selector != BaseHook.afterRemoveLiquidity.selector) {
+            revert Hooks.InvalidHookResponse();
+        }
+        uint256 nonzeroDeltaCount = manager.getNonzeroDeltaCount();
+        if (nonzeroDeltaCount == 0 && returnDelta == BalanceDeltaLibrary.ZERO_DELTA) {
+            return returnDelta;
+        }
+        if (
+            returnDelta.amount0() > int256(uint256(int256(delta.amount0())) * maxFeeBips / MAX_BIPS)
+                || returnDelta.amount1() > int256(uint256(int256(delta.amount1())) * maxFeeBips / MAX_BIPS)
+        ) {
```

Comment:
throw this in a library to make this code path more readable! and since you are doing this calculation twice.. maybe something like:

feeTake0, feeTaken1 = returnDelta.calculateFeesFrom(delta)
if (maxFee0 > feeTaken0 || maxFee1 > feeTaken1) revert HookTookTooMuchFee();